stock_move_line_removal_date/models/stock_move_line.py

```diff
@@ -0,0 +1,43 @@
+# Copyright 2020 Quartile Limited
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from datetime import datetime
+
+import pytz
+from dateutil.relativedelta import relativedelta
+from odoo import SUPERUSER_ID, fields, models
+
+
+class StockMoveLine(models.Model):
+    _inherit = "stock.move.line"
+
+    removal_date = fields.Date("Removal Date")
+
+    def _action_done(self):
+        res = super(StockMoveLine, self)._action_done()
+        for ml in self:
+            if ml.picking_id.picking_type_code == "incoming" and ml.removal_date:
+                removal_date = fields.Datetime.to_string(
+                    ml.convert_date_to_datetime(ml.sudo().removal_date, 12)
+                )
+                ml.lot_id.write({"removal_date": removal_date})
+        return res
+
+    def convert_date_to_datetime(self, date, hour):
+        # FIXME This method should ideally be put into another module
+        # dedicated to itself, adding to the Date class.
+        # It is also not optimal in terms of the performance to run this for
+        # every move line (should the diff be kept in the user record of the
+        # superuser? - in which case adjustments for DST is not possible).
+        dt = datetime(year=date.year, month=date.month, day=date.day, hour=hour,)
+        utc_timestamp = pytz.utc.localize(dt, is_dst=False)
+        su_tz = self.env["res.users"].browse(SUPERUSER_ID).tz or "UTC"
+        su_timestamp = pytz.timezone(su_tz).localize(dt).astimezone(pytz.utc)
+        # Following conditional branching is needed due to peculiar behavior
+        # of how the diff is calculated depending on the timezone.
+        if utc_timestamp >= su_timestamp:
+            diff = (utc_timestamp - su_timestamp).seconds / 3600
+            return utc_timestamp - relativedelta(hours=diff)
+        else:
+            diff = (su_timestamp - utc_timestamp).seconds / 3600
+            return utc_timestamp + relativedelta(hours=diff)
```

Comment:
How about we remove the if-statment, I believe we can pass negative values to the `hours`?
```
diff = (su_timestamp - utc_timestamp).seconds / 3600
return utc_timestamp + relativedelta(hours=diff)
```
`utc_timestamp - relativedelta(hours=diff)` should equal to `utc_timestamp + relativedelta(hours=-diff)`

Reply:
That unfortunately doesn't seem to work, and that's the "peculiar behavior" I mention in the comment.

For example, when timestamp_a (e.g. 12:00) < timestamp_b (e.g. 3:00), if you do timestamp_b minus timestamp_a, the result will be 15 hours instead of -9 hours. Maybe there is a better way to work around this, but I didn't get the time to do more research.

Reply:
I see, the calculation makes sense to me now.

Reply:
I think we can optimize code with this too. please confirm so I will add this too with test-case commit.

        diff = (su_timestamp - utc_timestamp).seconds / 3600
        if utc_timestamp >= su_timestamp:
            diff = (utc_timestamp - su_timestamp).seconds / 3600
        return utc_timestamp + relativedelta(hours=diff)